server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -2464,203 +2465,255 @@ public boolean assertAllListenersResolved() {
             } else {
                 updatedEntry = entry;
             }
             entries.add(updatedEntry);
         }
-        if (changedCount > 0) {
-            logger.trace("changed cluster state triggered by [{}] snapshot state updates and resulted in starting " +
-                    "[{}] shard snapshots", changedCount, startedCount);
-            return ClusterStateTaskExecutor.ClusterTasksResult.<ShardSnapshotUpdate>builder().successes(tasks)
-                    .build(ClusterState.builder(currentState).putCustom(SnapshotsInProgress.TYPE,
-                            SnapshotsInProgress.of(entries)).build());
+
+        private ImmutableOpenMap.Builder<RepositoryShardId, ShardSnapshotStatus> clonesBuilder() {
+            assert shardsBuilder == null;
+            if (clonesBuilder == null) {
+                clonesBuilder = ImmutableOpenMap.builder(current.clones());
+            }
+            return clonesBuilder;
         }
-        return ClusterStateTaskExecutor.ClusterTasksResult.<ShardSnapshotUpdate>builder().successes(tasks).build(currentState);
-    };
+
+        private ImmutableOpenMap.Builder<ShardId, ShardSnapshotStatus> shardsBuilder() {
+            assert clonesBuilder == null;
+            if (shardsBuilder == null) {
+                shardsBuilder = ImmutableOpenMap.builder(current.shards());
+            }
+            return shardsBuilder;
+        }
+
+        /**
+         * Start shard level operation for given {@code shardId}.
+         *
+         * @param newStates   builder for updated shard states mapping
+         * @param nodeId      node id to execute started operation on
+         * @param generation  shard generation to base started operation on
+         * @param shardId     shard identifier of shard to start operation for
+         */
+        private <T> void startShard(ImmutableOpenMap.Builder<T, ShardSnapshotStatus> newStates, String nodeId, String generation,
+                                    T shardId) {
+            startShard(newStates, shardId, new ShardSnapshotStatus(nodeId, generation));
+        }
+
+        /**
+         * Start shard level operation for given {@code shardId}.
+         *
+         * @param newStates builder for updated shard states mapping
+         * @param shardId   shard identifier of shard to start operation for
+         * @param newState  new shard task state for operation to start
+         */
+        private <T> void startShard(ImmutableOpenMap.Builder<T, ShardSnapshotStatus> newStates, T shardId, ShardSnapshotStatus newState) {
+            logger.trace("[{}] Starting [{}] on [{}] with generation [{}]", current.snapshot(), shardId, newState.nodeId(),
+                    newState.generation());
+            newStates.put(shardId, newState);
+            currentTaskIterator.remove();
+            startedCount++;
+        }
+
+        private <T> void updateExistingShardTask(ImmutableOpenMap<T, ShardSnapshotStatus> existingStates,
```

Comment:
Maybe `executeShardSnapshotUpdate`? Also I reckon it'd be nicer to just duplicate this method for the two cases (snapshots and clones) rather than passing `existingStates` and `newStates` in as arguments.

Reply:
++ Renamed

But I'd rather not duplicate the method. I have a follow-up in the pipeline that will shortly do away with `<T>` anyway :) Keeping it dry like this makes the follow-up diff a lot clearer I hope (+ tbh. I like it better this way even in isolation :)).

Reply:
acked